File: scratch/analyze.js
const fs = require('fs');

function analyze(filepath) {
    const content = fs.readFileSync(filepath, 'utf8');
    const lines = content.split('\n');
    
    let innerHtmlCount = 0;
    let listenerCount = 0;
    let snapshotCount = 0;
    let renderCount = 0;
    
    const issues = [];
    
    lines.forEach((line, i) => {
        const lineNum = i + 1;
        
        if (line.includes('.innerHTML =') || line.includes('.innerHTML +=')) {
            innerHtmlCount++;
        }
        
        if (line.includes('.addEventListener(')) {
            listenerCount++;
            if (line.match(/\.addEventListener\(['"](scroll|resize)['"]/)) {
                if (!line.includes('throttle') && !line.includes('debounce')) {
                    issues.push(`Line ${lineNum}: Potential unthrottled scroll/resize listener -> ${line.trim()}`);
                }
            }
        }
        
        if (line.includes('onSnapshot(')) {
            snapshotCount++;
        }
        
        if (line.match(/function\s+\w*render/i) || line.match(/render\w*\s*=/i)) {
            renderCount++;
        }
        
        // Check for multiple DOM queries in loops
        if (line.match(/document\.query/)) {
            // this is a naive check just to see where they are
        }
    });
    
    console.log(`Total innerHTML assignments: ${innerHtmlCount}`);
    console.log(`Total event listeners added: ${listenerCount}`);
    console.log(`Total Firestore snapshots: ${snapshotCount}`);
    console.log(`Total render functions: ${renderCount}`);
    
    issues.slice(0, 20).forEach(issue => console.log(issue));
}

analyze('app.js');

File: scratch/analyze.py
import re
import sys

def analyze(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    issues = []
    
    for i, line in enumerate(lines):
        # Unoptimized window listeners
        if re.search(r'window\.addEventListener\([\'"](scroll|resize)[\'"]', line):
            if 'throttle' not in line and 'debounce' not in line:
                issues.append((i+1, "Potential unthrottled scroll/resize listener", line.strip()))
        
        # innerHTML in loops (rough heuristic)
        if 'innerHTML +=' in line:
            issues.append((i+1, "innerHTML += is slow, use documentFragment or insertAdjacentHTML", line.strip()))
            
        # Multiple DOM selections that could be cached (rough heuristic - just flag all querySelector inside for/while)
        if re.search(r'(for\s*\(|while\s*\()', line):
            # This is too complex for simple regex, but let's check for inner querySelector
            pass
            
    # Check for innerHTML usage count
    inner_html_count = sum(1 for line in lines if '.innerHTML =' in line or '.innerHTML +=' in line)
    
    # Check for event listeners count
    listener_count = sum(1 for line in lines if '.addEventListener(' in line)
    
    # Check for Firestore snapshot
    snapshot_count = sum(1 for line in lines if 'onSnapshot(' in line)
    
    print(f"Total innerHTML assignments: {inner_html_count}")
    print(f"Total event listeners added: {listener_count}")
    print(f"Total Firestore snapshots: {snapshot_count}")
    
    for issue in issues[:20]:
        print(f"Line {issue[0]}: {issue[1]} -> {issue[2]}")

if __name__ == '__main__':
    analyze('app.js')
